### PR TITLE
composite-checkout: Add WeChat Pay

### PR DESCRIPTION
**Pull request by sirbrillig**
#### Changes proposed in this Pull Request

This adds the "WeChat Pay" payment method to composite-checkout.

#### Screenshots

Old checkout:

Screen Shot 2020-06-29 at 12 53 18 PM

wechat-old-qr

New checkout:

wechat-new-open

wechat-new-qr

#### Testing instructions

- Apply D45650-code to force WeChat Pay to be available.
- Sandbox the store.
- Add a plan to your cart and visit checkout. 
- Add `?flags=composite-checkout-force` to the query string in the URL and reload the page because currently it is disabled for international currencies.
- Verify that "WeChat Pay" is an option displayed in the final payment step.
- Verify that you cannot pay if the "Name" field is empty.
- Fill out the "Name" field and submit the payment.
- Verify that you see a QR code appear and it is scrolled into view.
- In a *new window or tab* open the link at the bottom that reads "To open and pay with the wechat pay app directly, click here".
- In the new window, you should see a Stripe test page.
- Click to accept the payment and verify that the payment completes successfully in the original window and that the plan is purchased.

### CODE REVIEW

File: client/my-sites/checkout/composite-checkout/wechat.js [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+		<WeChatFormWrapper>
+			<WeChatField
+				id="cardholderName"
+				type="Text"
+				autoComplete="cc-name"
+				label={ __( 'Your name' ) }
+				value={ customerName?.value ?? '' }
+				onChange={ changeCustomerName }
+				isError={ customerName?.isTouched && customerName?.value.length < 3 }
+				errorMessage={ __( 'Your name must contain at least 3 characters' ) }
```
Comment (a8ci18n) [inline client/my-sites/checkout/composite-checkout/wechat.js:104]: :information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **21** times:
`translate( 'Your Name must contain at least 3 characters' )` *ES Score: 9*

**Comment by fditrapani** [comment]
I was having trouble getting the branch running locally so I updated the code for the logo and pushed the updates. Let me know if that fixes the spacing in the screenshot.

**Comment by nsakaimbo** [comment]
@sirbrillig Seeing this build error for the wp-desktop source build:

```
ModuleNotFoundError: Module not found: Error: Can't resolve './cart-provider' in '/home/circleci/wp-desktop/calypso/client/my-sites/checkout/composite-checkout'
    at /home/circleci/wp-desktop/calypso/node_modules/webpack/lib/Compilation.js:925:10
    at /home/circleci/wp-desktop/calypso/node_modules/webpack/lib/NormalModuleFactory.js:401:22
 ```

[link](https://app.circleci.com/pipelines/github/Automattic/wp-desktop/18142/workflows/6a8960a9-8b3e-40ad-9430-94dc94a7e436/jobs/66551) (under `Build sources`) step

**Comment by sirbrillig** [comment]
Oops! It worked locally but I had forgotten to add a file. Sorry about that!

File: client/my-sites/checkout/composite-checkout/payment-methods/wechat.js [hunk 2]
```diff
@@ -0,0 +1,304 @@
...
+		<WeChatFormWrapper>
+			<WeChatField
+				id="cardholderName"
+				type="Text"
+				autoComplete="cc-name"
+				label={ __( 'Your name' ) }
+				value={ customerName?.value ?? '' }
+				onChange={ changeCustomerName }
+				isError={ customerName?.isTouched && customerName?.value.length < 3 }
+				errorMessage={ __( 'Your name must contain at least 3 characters' ) }
```
Comment (a8ci18n) [inline client/my-sites/checkout/composite-checkout/payment-methods/wechat.js:107]: :information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **21** times:
`translate( 'Your Name must contain at least 3 characters' )` *ES Score: 9*

**Comment by nbloomf** [comment]
I tested this by manually setting my geo to CN and sandboxing the API. The successful purchase worked as expected, thank you page loaded, products provisioned, and receipt generated. There was one console error which I do not think is related to this PR (but may be related to how we use the history in checkout):

image

There do appear to be some legit bugs in the payment failure flow. In this case the pending payment on the checkout page updated as expected, but checkout got stuck in a "processing" state. Also, and this is minor, but the grayed-out styling of the pending state isn't applied to all the components in the payment method.

image

**Comment by sirbrillig** [comment]
> There do appear to be some legit bugs in the payment failure flow. In this case the pending payment on the checkout page updated as expected, but checkout got stuck in a "processing" state

I ran into this as well during my testing, but I went away and when I came back it had reported a failure, so I assumed it just takes a while. However, maybe something isn't working right there. I'll check again.

**Comment by sirbrillig** [comment]
There _was_ a bug. Resetting the transaction status did not also reset the form status. I've made that automatic now.

**Comment by nbloomf** [comment]
Looks good now- both successful and failed purchases work as expected.

**Comment by sirbrillig** [comment]
Oooh... tests are broken. Good opportunity to add some tests for the new transaction behavior though.

File: client/my-sites/checkout/composite-checkout/payment-methods/wechat.js (hunk 2 above)
Comment (sirbrillig) [reply client/my-sites/checkout/composite-checkout/payment-methods/wechat.js:107]: Just a capitalization change. The original was incorrect.

**Comment by a8ci18n** [comment]
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/3988656

Thank you @sirbrillig for including a screenshot in the description! This is really helpful for our translators.

**Comment by a8ci18n** [comment]
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/3988656).

**Pull request merged**